sdk/core/azure-core/src/main/java/com/azure/core/util/polling/implementation/LongRunningOperation.java

```diff
@@ -0,0 +1,212 @@
+    }
+
+    /**
+     * @return a Flux, upon subscription starts the long running operation followed by polling.
+     */
+    public Flux<AsyncPollResponse<PollResultT, FinalResultT>> begin() {
+        // Get a Flux on which ONLY ONE subscription can start LRO successfully and ALL subscriptions
+        // can poll. Polling is not shared, each subscription starts it's own polling, if user need
+        // multi-casting for any reason then they should call share() on returned Flux.
+        return
+        this.lroStartMonoWithResponseCache
+            .flatMapMany(lroStartResponse -> pollingLoop(lroStartResponse));
+    }
+
+    /**
+     * Returns a decorated Mono, upon subscription it internally subscribes to the Mono that starts LRO.
+     * The decorated Mono caches the response of LRO initiate call, this cached response will be replayed
+     * for any future subsequent subscriptions.
+     *
+     * Note: we can't use the reactor standard cache() operator, because it caches error terminal signal
+     * and forward it to any future subscriptions. If there is an error from lroStartMono
+     * then we don't want to cache it but just forward it to subscription that initiated the failed LRO.
+     * For any future subscriptions we don't want to forward the past error instead lroStartMono
+     * should again invoked. Once a subscription manage to get a successful event from lroStartMono call 
+     * then we want to cache it and replay it to any future subscriptions.
+     *
+     * The decorated Mono also handles concurrent subscription to it, on such event only one of them
+     * will be able to call lroStartMono and other subscriptions will keep resubscribing until
+     * it sees a cached response or get a chance to call lroStartMono as the one previously
+     * entered the critical section got an error from lroStartMono.
+     *
+     * @param lroStartMono the Mono upon subscription initiate/start the long running operation.
+     * @return
+     */
+    private Mono<PollResponse<PollResultT>> lroStartMonoWithResponseCache(final Mono<PollResultT> lroStartMono) {
+        return Mono.defer(() -> {
+            if (this.lroStartResponse != null) {
+                return Mono.just(this.lroStartResponse);
+            }
+            if (this.guardLroStartCall.compareAndSet(this, 0, 1)) {
+                return lroStartMono.map((PollResultT response) -> {
+                    this.lroStartResponse = new PollResponse<>(PollResponse.OperationStatus.NOT_STARTED, response);
+                    return this.lroStartResponse;
+                })
+                .switchIfEmpty(Mono.error(new RuntimeException("lroStartMono should produce a Response.")))
+                .doOnError(throwable -> guardLroStartCall.compareAndSet(this, 1, 0));
+            } else {
+                return Mono.empty();
+            }
+        }).repeatWhenEmpty((Flux<Long> longFlux) -> longFlux.concatMap(ignored -> Flux.just(true)));
```

Comment:
Is the call to `repeatWhenEmpty` here for threads which failed to acquire the lock to wait for the 1st thread to complete the work?

Reply:
Yes, that is correct. we do busy loop here instead of blocking to handle very rare but still possible concurrency bug

Reply:
And we want only one successful call to happen on start lro